### DISCUSSION TITLE
How to create In-step attachments

### DISCUSSION THREAD

**Discussion by kiragengis** [allure-framework/allure2#2395] [Questions & Support] [answered]
With the help of @delatrie I have created my logging system using Allure. Now I am facing the challenge that I want to add screenshots and formatted text like in this report:
Formatted text: https://allure-framework.github.io/allure-demo/9/#suites/a2891ce60e520f56ae25e6caf68ea773/66a85fad08b7c91b/
Screenshot: https://allure-framework.github.io/allure-demo/9/#suites/2d326f1b8f98beb0493c8c9556b35a3b/1510e2e7a4f5cdd4/

So far ,I can only add attachments at the end of the report. I am almost sure this is because I am capturing the screenshots in` Kind.Before` and `Kind.After` like this:

```
[Advice(Kind.Before)]
public void AddScreenshotBefore(
    [Argument(Source.Instance)] object instance,
    [Argument(Source.Name)] string name)
{
    TakeScreenshotForLogs(instance, name);
}

private void TakeScreenshotForLogs(object instance, string name)
{
    PageObject element = instance as PageObject;
    byte[] screenshotBytes = element.GetScreenshot();
    string imageTitle = element.Name != "" ? element.Name : "PageObject without name";
    AllureApi.AddAttachment($"{imageTitle}.png", "image/png", screenshotBytes);
}
```
How can I add the screenshot as part of my step and not to add a sub step o new step for this?

**Comment by delatrie** [comment, marked as answer]
Adding an attachment to a step requires the step's context to be active. That can't be done with `[Advice(Kind.After)]`. You have to use `[Advice(Kind.Around)]` instead. The primary headache here is handling all cases involving async step functions. There are 6 cases in total, depending on the return type:

  - `void`
  - some non-task value like `string`
  - `Task`
  - `Task<T>`
  - `ValueTask`
  - `ValueTask<T>`

That would cover most cases (but not 100% since there are other things you can await beside those four task types, see [here](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#12982-awaitable-expressions) for more details).

> [!NOTE]
> Functions returning `ValueTask` and `ValueTasl<T>` aren't supported by `[AllureStep]`, see allure-framework/allure-csharp#456

I'll show you the skeleton code you can use to complete the solution. The code allows using non-async steps and steps that return `Task<T>`. You can support other cases by adding more `if-else` branches.

```c#
[Aspect(Scope.Global)]
public class AddScreenshotsAspect
{
    [Advice(Kind.Around)]
    public object AddReturnValueToAllureStep(
        [Argument(Source.Instance)] object? instance,
        [Argument(Source.ReturnType)] Type returnType,
        [Argument(Source.Arguments)] object[] arguments,
        [Argument(Source.Target)] Func<object[], object> target
    )
    {
        if (returnType.IsConstructedGenericType && returnType.GetGenericTypeDefinition() == typeof(Task<>))
        {
            return typeof(AddScreenshotsAspect)
                .GetMethod(
                    nameof(AddScreenshotsTaskWithReturn),
                    System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Static
                )!
                .MakeGenericMethod(
                    returnType.GetGenericArguments()[0]
                )
                .Invoke(null, new[] { instance, target, arguments })!;
        }

        /* handle other cases in else if branches */

        /* the default case (void and non-task return types) */
        var returnValue = target(arguments);
        this.AddScreenshot(instance);
        return returnValue;
    }

    static async Task<T> AddScreenshotsTaskWithReturn<T>(
        PageObject page,
        Func<object[], object> invokeAsyncStep,
        object[] arguments
    )
    {
        var returnValue = await ((Task<T>)invokeAsyncStep(arguments)).ConfigureAwait(false); // you might not need ConfigureAwait depending on the environment
        this.AddScreenshot(instance);
        return returnValue;
    }

/* ... */
}
```

I didn't show the `AddScreenshot` function that actually contains the `AllureApi.AddAttachment` call. It should be straightforward.

As you can see, if you only want to support synchronous step functions, the code will be much simpler.

The only thing that remains is to ensure the advice that creates and resets a step's context is executed before the code above. You have to overwrite the priority of the corresponding aspect to do that:

```c#
[Injection(typeof(AddScreenshotsAspect), Priority = 0)]
[Injection(typeof(Allure.Net.Commons.Steps.AllureStepAspectBase), Priority = 1)]
public class StepWithScreenshotAttribute : AllureStepAttribute
{
    public StepWithScreenshotAttribute(string? name = null) : base(name) { }
}
```

**Comment by kiragengis** [comment]
Thank you @delatrie ! You are always helping me :) I owe you at least a cup of coffee.
